Improve test-classpath configuration for split environment

## Context
I recently wanted to add unit tests to my mod.
In the project, loom is configured with `splitEnvironmentSourceSets()`.

When trying to run the gradle 'test' task, I get the following error:
```
Minecraft game provider couldn't locate the game! The game may be absent from the class path, lacks some expected files, suffers from jar corruption or is of an unsupported variety/version.
```

The issue is, that the game common and environment jars are not in the class path.
I fixed the issue by manually adding this to my buildscript:
```gradle
configurations {
    // common
    testCompileClasspath.extendsFrom(minecraftCommonNamedCompile)
    testRuntimeClasspath.extendsFrom(minecraftCommonNamedRuntime)
    // client only
    testCompileClasspath.extendsFrom(minecraftClientOnlyNamedCompile)
    testRuntimeClasspath.extendsFrom(minecraftClientOnlyNamedRuntime)
}
```

Now, game provider works, but this error occurs instead:
```
...
Caused by: java.lang.IllegalArgumentException: The specified resource 'fabric-networking-api-v1.client.mixins.json' was invalid or could not be read
	at org.spongepowered.asm.mixin.transformer.MixinConfig.create(MixinConfig.java:1290)
	at org.spongepowered.asm.mixin.transformer.Config.create(Config.java:148)
	... 41 more
```
The client mixin files cannot be found (the actual file is different sometimes).
I fixed this by adding this to my `build.gradle`:
```gradle
configurations {
    testRuntimeClasspath.extendsFrom(modRuntimeClasspathClientMapped)
}
```

Then, everything is working as expected for my client+server project.

## For server-only projects
One of my other mods is configured with `serverOnlyMinecraftJar()`.
When executing the unit tests, the same game provider error occurs.
I fixed it with:
```gradle
configurations {
    testCompileClasspath.extendsFrom(minecraftCommonNamedCompile)
    testRuntimeClasspath.extendsFrom(minecraftCommonNamedRuntime)
}
```
However, the error with the mixins does not occur.

## Issue
This manual classpath configuration is a little confusing for mod developers who want to test their code with unit tests, especially when they don't know Gradle well...

Maybe loom could automatically do this configuration for projects with `splitEnvironmentSourceSets()` or `serverOnlyMinecraftJar()`?
I am pretty sure that `clientOnlyMinecraftJar()` needs extra configuration too.
At the very least, I think this should be documented somewhere.

Cheers

--- a/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftSourceSets.java
+++ b/src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftSourceSets.java
@@ -205,6 +205,12 @@ public abstract sealed class MinecraftSourceSets permits MinecraftSourceSets.Sin
 			extendsFrom(project, clientOnlySourceSet.getCompileClasspathConfigurationName(), mainSourceSet.getCompileClasspathConfigurationName());
 			extendsFrom(project, clientOnlySourceSet.getRuntimeClasspathConfigurationName(), mainSourceSet.getRuntimeClasspathConfigurationName());
 
+			// Test source set depends on client
+			final SourceSet testSourceSet = SourceSetHelper.getSourceSetByName(SourceSet.TEST_SOURCE_SET_NAME, project);
+			extendsFrom(project, testSourceSet.getCompileClasspathConfigurationName(), clientOnlySourceSet.getCompileClasspathConfigurationName());
+			extendsFrom(project, testSourceSet.getRuntimeClasspathConfigurationName(), clientOnlySourceSet.getRuntimeClasspathConfigurationName());
+			project.getDependencies().add(testSourceSet.getImplementationConfigurationName(), clientOnlySourceSet.getOutput());
+
 			RemapConfigurations.configureClientConfigurations(project, clientOnlySourceSet);
 
 			// Include the client only output in the jars

--- a/src/test/groovy/net/fabricmc/loom/test/integration/SplitProjectTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/SplitProjectTest.groovy
@@ -43,6 +43,7 @@ class SplitProjectTest extends Specification implements GradleProjectTestTrait {
 
 		then:
 		result.task(":build").outcome == SUCCESS
+		result.task(":test").outcome == SUCCESS
 
 		where:
 		version << STANDARD_TEST_VERSIONS

--- a/src/test/groovy/net/fabricmc/loom/test/util/ExtractTestProject.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/util/ExtractTestProject.groovy
@@ -1,0 +1,83 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2024 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.test.util
+
+// A helper script to extract the test project from the resources and add the gradle wrapper
+// Useful for testing an intergration test project in a real environment
+class ExtractTestProject {
+	static void main(String[] args) {
+		if (args.length != 1) {
+			throw new IllegalArgumentException("Expected one argument: the project name")
+		}
+
+		def projectName = args[0]
+		def targetDirectory = new File("test/$projectName")
+		def sourceDirectory = new File("src/test/resources/projects/$projectName")
+
+		if (targetDirectory.exists()) {
+			targetDirectory.deleteDir()
+		}
+
+		copyDir(sourceDirectory, targetDirectory)
+		copyDir(new File("gradle"), new File(targetDirectory, "gradle"))
+		copyFile(new File("gradlew"), new File(targetDirectory, "gradlew"))
+		copyFile(new File("gradlew.bat"), new File(targetDirectory, "gradlew.bat"))
+
+		new File(targetDirectory, "settings.gradle").text = """
+            pluginManagement {
+                repositories {
+                    maven {
+                        name = 'Fabric'
+                        url = 'https://maven.fabricmc.net/'
+                    }
+                    mavenCentral()
+                    gradlePluginPortal()
+                    mavenLocal()
+                }
+            }
+            """
+
+		def buildGradle = new File(targetDirectory, "build.gradle")
+		buildGradle.text = buildGradle.text.replace("id 'fabric-loom'", "id 'fabric-loom' version '1.8.local'")
+	}
+
+	private static void copyDir(File source, File target) {
+		source.eachFileRecurse { file ->
+			if (file.isDirectory()) {
+				return
+			}
+
+			def relativePath = source.toPath().relativize(file.toPath())
+			def targetFile = new File(target, relativePath.toString())
+			copyFile(file, targetFile)
+		}
+	}
+
+	private static void copyFile(File source, File target) {
+		target.parentFile.mkdirs()
+		target.bytes = source.bytes
+		println("Copied $source to $target")
+	}
+}

--- a/src/test/resources/projects/splitSources/build.gradle
+++ b/src/test/resources/projects/splitSources/build.gradle
@@ -1,5 +1,5 @@
 plugins {
-	id 'fabric-loom' version '0.12.local'
+	id 'fabric-loom'
 	id 'maven-publish'
 }
 
@@ -24,11 +24,18 @@ loom {
 dependencies {
 	minecraft "com.mojang:minecraft:1.18.2"
 	mappings "net.fabricmc:yarn:1.18.2+build.1:v2"
-	modImplementation "net.fabricmc:fabric-loader:0.13.3"
+	modImplementation "net.fabricmc:fabric-loader:0.16.5"
+	testImplementation "net.fabricmc:fabric-loader-junit:0.16.5"
 
 	modImplementation "net.fabricmc.fabric-api:fabric-api:0.47.8+1.18.2"
 
 	shade "com.googlecode.json-simple:json-simple:1.1.1"
+
+//	testImplementation sourceSets.client.output
+}
+
+test {
+	useJUnitPlatform()
 }
 
 tasks.withType(JavaCompile).configureEach {

--- a/src/test/resources/projects/splitSources/build.gradle
+++ b/src/test/resources/projects/splitSources/build.gradle
@@ -30,8 +30,6 @@ dependencies {
 	modImplementation "net.fabricmc.fabric-api:fabric-api:0.47.8+1.18.2"
 
 	shade "com.googlecode.json-simple:json-simple:1.1.1"
-
-//	testImplementation sourceSets.client.output
 }
 
 test {

--- a/src/test/resources/projects/splitSources/src/client/resources/modid.client.mixins.json
+++ b/src/test/resources/projects/splitSources/src/client/resources/modid.client.mixins.json
@@ -1,7 +1,7 @@
 {
   "required": true,
   "minVersion": "0.8",
-  "package": "net.fabricmc.example.mixin",
+  "package": "net.fabricmc.example.client.mixin",
   "compatibilityLevel": "JAVA_17",
   "mixins": [
   ],

--- a/src/test/resources/projects/splitSources/src/test/java/net/fabricmc/example/test/ExampleModTest.java
+++ b/src/test/resources/projects/splitSources/src/test/java/net/fabricmc/example/test/ExampleModTest.java
@@ -1,0 +1,17 @@
+package net.fabricmc.example.test;
+
+import net.fabricmc.example.client.ExampleModClient;
+
+import net.minecraft.client.MinecraftClient;
+import org.junit.jupiter.api.Test;
+
+public class ExampleModTest {
+	@Test
+	void testClientClass() {
+		// Check we can compile against our own client code
+		ExampleModClient exampleModClient = new ExampleModClient();
+
+		// And minecrafts client code
+		MinecraftClient client = null;
+	}
+}

